modules/contracts/apps/challenge/src/handlers/execute.rs

```diff
@@ -0,0 +1,497 @@
+
+    for friend in &friends {
+        if existing_friends.iter().any(|f| f.address == friend.address) {
+            return Err(AppError::AlreadyAdded {});
+        }
+    }
+
+    // validate the String addresses and convert them to Addr
+    // before saving
+    let friends: Vec<Friend<Addr>> = friends
+        .into_iter()
+        .map(|friend| friend.check(deps.as_ref()).unwrap())
+        .collect();
+
+    existing_friends.extend(friends);
+    CHALLENGE_FRIENDS.save(deps.storage, challenge_id, &existing_friends)?;
+
+    Ok(app.tag_response(
+        Response::new().add_attribute("challenge_id", challenge_id.to_string()),
+        "add_friends",
+    ))
+}
+
+pub fn remove_friends_from_challenge(
+    deps: DepsMut,
+    info: MessageInfo,
+    app: &ChallengeApp,
+    challenge_id: u64,
+    friend_addresses: Vec<Friend<String>>,
+) -> AppResult {
+    // Ensure the caller is an admin
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+
+    let mut existing_friends = CHALLENGE_FRIENDS
+        .may_load(deps.storage, challenge_id)?
+        .unwrap_or_default();
+
+    for friend in &friend_addresses {
+        if !existing_friends.iter().any(|f| f.address == friend.address) {
+            return Err(AppError::Std(StdError::generic_err(
+                "Friend not found for this challenge",
+            )));
+        }
+    }
+
+    existing_friends.retain(|f| {
+        !friend_addresses
+            .iter()
+            .any(|friend| f.address == friend.address)
+    });
+    CHALLENGE_FRIENDS.save(deps.storage, challenge_id, &existing_friends)?;
+    Ok(app.tag_response(
+        Response::new().add_attribute("challenge_id", challenge_id.to_string()),
+        "remove_friends",
+    ))
+}
+
+fn daily_check_in(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    app: &ChallengeApp,
+    challenge_id: u64,
+    metadata: Option<String>,
+) -> AppResult {
+    app.admin.assert_admin(deps.as_ref(), &info.sender)?;
+    let mut challenge = CHALLENGE_LIST.load(deps.storage, challenge_id)?;
+
+    // If the challenge has ended, we set the status to Over and return
+    if env.block.time > challenge.end {
+        match challenge.status {
+            ChallengeStatus::Active => {
+                challenge.status = ChallengeStatus::Over;
+                CHALLENGE_LIST.save(deps.storage, challenge_id, &challenge)?;
+                return Ok(app.tag_response(
+                    Response::new().add_attribute(
+                        "message",
+                        "Challenge has ended. You can no longer register a daily check in.",
+                    ),
+                    "daily check in",
+                ));
+            }
+            _ => {
+                return Err(AppError::Std(StdError::generic_err(format!(
+                    "Challenge has ended. Challenge end_timestamp is {:?} current timestamp is {}",
+                    challenge.end.seconds(),
+                    env.block.time.seconds()
```

Comment:
One of them is debug output second isn't :eyes:

Reply:
whoops, nice catch.